src/handlers/register.rs

```diff
@@ -32,24 +34,89 @@ pub async fn handler(
     claims.verify_basic(&state.auth_aud, None)?;
     let client_id = ClientId::from(claims.iss);
 
+    increment_counter!(state.metrics, register);
+
+    if let Some(tags) = body.tags {
+        increment_counter!(state.metrics, registration_overwrite);
+        overwrite_registration(&state, client_id.clone(), tags, body.relay_url).await?;
+    } else {
+        increment_counter!(state.metrics, registration_update);
+        update_registration(
+            &state,
+            client_id.clone(),
+            body.append_tags,
+            body.remove_tags,
+            body.relay_url,
+        )
+        .await?;
+    }
+
+    Ok(Response::default())
+}
+
+async fn overwrite_registration(
+    state: &Arc<AppState>,
+    client_id: ClientId,
+    tags: Vec<Arc<str>>,
+    relay_url: Arc<str>,
+) -> error::Result<Response> {
     state
         .registration_store
         .upsert_registration(
             client_id.value(),
-            body.tags.iter().map(AsRef::as_ref).collect(),
-            body.relay_url.as_ref(),
+            tags.iter().map(AsRef::as_ref).collect(),
+            relay_url.as_ref(),
         )
         .await?;
 
     state
         .registration_cache
         .insert(client_id.into_value(), CachedRegistration {
-            tags: body.tags.clone(),
-            relay_url: body.relay_url.clone(),
+            tags,
+            relay_url,
         })
         .await;
 
-    increment_counter!(state.metrics, register);
-
     Ok(Response::default())
 }
+
+async fn update_registration(
+    state: &Arc<AppState>,
+    client_id: ClientId,
+    append_tags: Option<Vec<Arc<str>>>,
+    remove_tags: Option<Vec<Arc<str>>>,
+    relay_url: Arc<str>,
+) -> error::Result<Response> {
+    if let (Some(append_tags), Some(remove_tags)) = (append_tags.clone(), remove_tags.clone()) {
+        for tag in remove_tags.iter() {
+            if append_tags.contains(tag) {
+                return Err(Error::InvalidUpdateRequest);
+            }
+        }
+    }
+
+    let registration = state
+        .registration_store
+        .get_registration(client_id.as_ref())
+        .await?;
+
+    let mut tags = registration.tags;
+
+    if let Some(remove_tags) = remove_tags {
+        for r_tag in remove_tags.iter() {
+            if let Some(index) = tags.iter().position(|tag| tag.eq(r_tag)) {
+                tags.remove(index);
+            }
+        }
+    }
+
+    if let Some(append_tags) = append_tags {
+        for a_tag in append_tags.iter() {
+            if !tags.contains(a_tag) {
+                tags.push(a_tag.clone());
+            }
+        }
+    }
```

Comment:
Would a HashSet be better?

```suggestion
    let new_tags = registration.tags.into_iter().collect::<HashSet<_>>()
        .difference(&remove_tags.unwrap_or_else(|| HashSet::new())).collect::<HashSet<_>>()
        .union(&append_tags.unwrap_or_else(|| HashSet::new())).collect::<HashSet<_>>()
        .into_iter().collect::<Vec<_>>();
```

Reply:
This one doesn't compile. A form that actually compiles is this:

```rust
    let tags = registration
        .tags
        .into_iter()
        .collect::<HashSet<_>>()
        .difference(&remove_tags.unwrap_or_else(|| HashSet::new()))
        .map(|t| t.clone())
        .collect::<HashSet<_>>()
        .union(&append_tags.unwrap_or_else(|| HashSet::new()))
        .map(|t| t.clone())
        .collect::<HashSet<_>>()
        .into_iter()
        .collect::<Vec<_>>();
```

This is really verging on write-only code, and I'm not sure there are a lot of computational advantages to this form (especially given the expected size of the collections).

Reply:
My concern was more over readability/guaranteed set behavior than performance. Thought set operations would be easier to understand and easier to make sure the correct behavior is implemented.

Reply:
I think this is a slightly cleaned up version that compiles:

```
    let tags = registration
        .tags
        .into_iter()
        .collect::<HashSet<_>>()
        .difference(&remove_tags.unwrap_or_else(|| HashSet::new()))
        .cloned()
        .collect::<HashSet<_>>()
        .union(&append_tags.unwrap_or_else(|| HashSet::new()))
        .cloned()
        .collect::<Vec<_>>();
```

Reply:
Up to you. Just a suggestion

Reply:
Sure, and I understand that in some way it's more elegant and could be more efficient on bigger collections, but I don't think we'll ever get anything much bigger than 10 items, so I'd rather go with legibility 😀

Reply:
Again, performance wasn't my concern haha. Set operations are more legible IMO than loops and if statements